### ISSUE TITLE
fit_generator - No change in accuracy

### ISSUE THREAD

**Issue by Shahnawazgrewal** [keras-team/keras#9423]
I modified a keras script 01 to train a resnet on my own dataset. It reads csv files of train and test sets. This script works perfectly however, I have MemoryError in case of large scale dataset. To resolve this situation I used fit_generator to work on batches to avoid MemoryError. Script 02 uses fit_generator to use train and test sets to train the resnet.
**Script 01:**

```
batch_size = 32
nb_classes = 4
nb_epoch = 10
data_augmentation = False

# input image dimensions
img_rows, img_cols = 256, 256
#  images are RGB.
img_channels = 3

img_lst_train = []
label_lst_train = []

img_lst_test = []
label_lst_test = []

print ('Reading training file')
with open('train-20news.csv') as csvfile:
    readCSV = csv.reader(csvfile, delimiter=',')
    for row in readCSV:
        print(row[0])
        img = cv2.imread(row[0])
        img_lst_train.append(img) 
        label_lst_train.append(row[1])

print ('Reading testing file')
with open('val-20news.csv') as csvfile:
    readCSV = csv.reader(csvfile, delimiter=',')
    for row in readCSV:
        #print(row[0])
        img = cv2.imread(row[0])
        img_lst_test.append(img) 
        label_lst_test.append(row[1])  

img_lst_train = np.array(img_lst_train)
label_lst_train = np.array(label_lst_train)
img_lst_test = np.array(img_lst_test)
label_lst_test = np.array(label_lst_test)

X_train = img_lst_train
y_train = label_lst_train
X_test  = img_lst_test
y_test  = label_lst_test

del img_lst_train 
del label_lst_train
del img_lst_test 
del label_lst_test

# Convert class vectors to binary class matrices.
Y_train = np_utils.to_categorical(y_train, nb_classes)
Y_test = np_utils.to_categorical(y_test, nb_classes)


X_train = X_train.astype('float32', copy=False)
X_test = X_test.astype('float32', copy=False)

# subtract mean and normalize
mean_image = np.mean(X_train, axis=0)
X_train -= mean_image
X_test -= mean_image
X_train /= 128.
X_test /= 128.

model = resnet.ResnetBuilder.build_resnet_18((img_channels, img_rows, img_cols), nb_classes)
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])

if not data_augmentation:
    print('Not using data augmentation.')
    model.fit(X_train, Y_train,
              batch_size=batch_size,
              nb_epoch=nb_epoch,
              validation_data=(X_test, Y_test),
              shuffle=True,
              callbacks=[lr_reducer, early_stopper, csv_logger])
else:
    print('Using real-time data augmentation.')
    # This will do preprocessing and realtime data augmentation:
    datagen = ImageDataGenerator(
        featurewise_center=False,  # set input mean to 0 over the dataset
        samplewise_center=False,  # set each sample mean to 0
        featurewise_std_normalization=False,  # divide inputs by std of the dataset
        samplewise_std_normalization=False,  # divide each input by its std
        zca_whitening=False,  # apply ZCA whitening
        rotation_range=0,  # randomly rotate images in the range (degrees, 0 to 180)
        width_shift_range=0.1,  # randomly shift images horizontally (fraction of total width)
        height_shift_range=0.1,  # randomly shift images vertically (fraction of total height)
        horizontal_flip=False, # randomly flip images
        vertical_flip=False)  # randomly flip images

    # Compute quantities required for featurewise normalization
    # (std, mean, and principal components if ZCA whitening is applied).
    datagen.fit(X_train)

    # Fit the model on the batches generated by datagen.flow().
    model.fit_generator(datagen.flow(X_train, Y_train, batch_size=batch_size),
                        steps_per_epoch=X_train.shape[0] // batch_size,
                        validation_data=(X_test, Y_test),
                        epochs=nb_epoch, verbose=1, max_q_size=100,
                        callbacks=[lr_reducer, early_stopper, csv_logger])
```
Modified script to resolve MemoryError and is based on fit_generator

**Script 02:**
```
def read_avg(source):
    with open(source) as csvfile:
        readCSV = csv.reader(csvfile, delimiter=',')
        avg_sum = None
        number_of_record=0
        for row in readCSV:
            img = cv2.imread(row[0])
            img = img.astype('float32')
            if avg_sum is not None:
                avg_sum+=img
            else:
                avg_sum=img
            number_of_record+=1
            #print (number_of_record)
    return avg_sum/number_of_record


lr_reducer = ReduceLROnPlateau(factor=np.sqrt(0.1), cooldown=0, patience=5, min_lr=0.5e-6)
early_stopper = EarlyStopping(min_delta=0.001, patience=10)
csv_logger = CSVLogger('resnet18_cifar10.csv')

batch_size = 32
nb_classes = 4
nb_epoch = 10
data_augmentation = False

# input image dimensions
img_rows, img_cols = 256, 256
# images are RGB.
img_channels = 3


img_lst_train = []
label_lst_train = []
img_lst_test = []
label_lst_test = []

print ('Reading training file')
with open('train-20news.csv') as csvfile:
    readCSV = csv.reader(csvfile, delimiter=',')
    for row in readCSV:
        print(row[0])
        img = cv2.imread(row[0])
        img_lst_train.append(img) 
        label_lst_train.append(row[1])

print ('Reading testing file')

with open('val-20news.csv') as csvfile:
    readCSV = csv.reader(csvfile, delimiter=',')
    for row in readCSV:
        #print(row[0])
        img = cv2.imread(row[0])
        img_lst_test.append(img) 
        label_lst_test.append(row[1])


img_lst_train = np.array(img_lst_train)
label_lst_train = np.array(label_lst_train)
img_lst_test = np.array(img_lst_test)
label_lst_test = np.array(label_lst_test)


X_train = img_lst_train
y_train = label_lst_train
X_test  = img_lst_test
y_test  = label_lst_test

# Convert class vectors to binary class matrices.
Y_train = np_utils.to_categorical(y_train, nb_classes)
Y_test = np_utils.to_categorical(y_test, nb_classes)

del img_lst_train 
del label_lst_train

del img_lst_test 
del label_lst_test

avg_image_train=read_avg("train-20news.csv")
avg_image_test=read_avg("val-20news.csv")


def subtract_mean_gen(x_source,y_source,avg_image,batch):
    batch_list_x=[]
    batch_list_y=[]
    while 1: #run forever, so you can generate elements indefinitely
        for line,y in zip(x_source,y_source):
            x=line.astype('float32')
            x=x-avg_image
            x /= 128.
            #x /= 128.
            batch_list_x.append(x)
            batch_list_y.append(y)
            if len(batch_list_x) == batch:
                yield (np.array(batch_list_x),np.array(batch_list_y))
                batch_list_x=[]
                batch_list_y=[]



model = resnet.ResnetBuilder.build_resnet_18((img_channels, img_rows, img_cols), nb_classes)
model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy'])


val = subtract_mean_gen(X_test,Y_test,avg_image_test,batch_size)
model.fit_generator(subtract_mean_gen(X_train,Y_train,avg_image_train,batch_size), steps_per_epoch=X_train.shape[0]//batch_size,epochs=nb_epoch, verbose=1, callbacks=[lr_reducer, early_stopper, csv_logger],
                    validation_data = subtract_mean_gen(X_test,Y_test,avg_image_test,batch_size),
                    validation_steps = X_test.shape[0]//batch_size)
```

With script 01, I obtained an accuracy of 98% however with script 02, I obtained an accuracy of 18% with each epoch. Please let me know if I am using fit_generator correctly

**Output of Script 02**
```
Epoch 1/10
249/249 [==============================] - 76s 304ms/step - loss: 1.1363 - categorical_accuracy: 0.8813 - val_loss: 13.3803 - val_categorical_accuracy: 0.1809
Epoch 2/10
249/249 [==============================] - 72s 291ms/step - loss: 1.4564 - categorical_accuracy: 0.6728 - val_loss: 3.5057 - val_categorical_accuracy: 0.1809
Epoch 3/10
249/249 [==============================] - 73s 291ms/step - loss: 1.2680 - categorical_accuracy: 0.6796 - val_loss: 2.5300 - val_categorical_accuracy: 0.1809
Epoch 4/10
249/249 [==============================] - 73s 293ms/step - loss: 1.4486 - categorical_accuracy: 0.6507 - val_loss: 3.8906 - val_categorical_accuracy: 0.1809
Epoch 5/10
249/249 [==============================] - 73s 295ms/step - loss: 1.1915 - categorical_accuracy: 0.7097 - val_loss: 3.3898 - val_categorical_accuracy: 0.1809

```
